### DISCUSSION TITLE
Permission denied when using service_role key

### DISCUSSION THREAD

**Discussion by zidkim** [supabase/supabase#11547] [Questions] [answered]
I'd like to be able to delete users from auth.users table programmatically from a client/server.

I understand that to update/delete from this table, I would need to use the service_role key when initializing the supabase client. 

Regardless of this, I keep getting "permission denied for table users".

I've reviewed other threads+google but have yet to find success.
I should also note that I've checked servicekey is accurate and that it is correctly passed into the server environment. 

Here's my setup:

1. Initialize supabase client with service_role key

```
const getServiceSupabase = () =>
  createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL,
    process.env.SUPABASE_SERVICE_KEY
  );

export { getServiceSupabase };
```

2. nextjs API routes

```
import { getServiceSupabase } from "@/lib/supabase";

export default async function handler(req, res) {
  const supabaseAdmin = getServiceSupabase();

  const { method } = req;
  const { userId } = req.body;

  const response = await supabaseAdmin.rpc("delete_supabase_user", {
    user_id: userId,
  });
  console.log("🚀 ~ handler ~ response", response);

  res.status(200).send({ status: "OK" });
}
```

This is the PostgreSQL command I made in the editor

```
create function delete_supabase_user(user_id uuid)
returns void
language plpgsql
as $$
begin
delete from auth.users where id = user_id;
end; $$;
```

And below is the resulting function in the dashboard.

![image](https://user-images.githubusercontent.com/46095969/211388050-e70e1057-367f-401d-8598-67a7572be296.png)

Please advise, and thank you for your help!

**Comment by GaryAustin1** [comment, marked as answer]
The API cannot access the auth schema.   You will need to use an rpc function and your Postgres function has to be "security definer" type to have Postgres user privileges.   Service_role is not enough.

**Comment by zidkim** [comment]
@GaryAustin1 Exactly what I needed, thank you. Still very new to SQL/PostgreSQL, so not too confident with the below code, but it seems to work. If there's a better/more secure way of defining this, please do humble me, and thanks @GaryAustin1!

```
create function delete_supabase_user_auth(user_id uuid)
returns void
language plpgsql
security definer
set search_path = public
as $$
begin
delete from auth.users where id = user_id;
end; $$;
```

**Reply by GaryAustin1** [reply]
I'm not sure just deleting a user from auth.users is really the way to go versus using supabase.auth.admin.deleteUser().   I was just showing you how to access auth schema.  There are other tables associated with a user signed in with Auth, including the identities table.   I do not know if cascade delete is setup and what other things might need to occur.  In general you should avoid messing with auth, and storage schemas directly.

**Reply by GaryAustin1** [reply]
Also your call pretty much means if I know a user's id I can delete it.  You don't check if the user calling the function is service_role or has permissions.  At a minimum you should use auth.uid() and not pass the id if you are trying to let a user do it.

I would say what you have now is not secure.

**Reply by zidkim** [reply]
Ah using `auth.uid()` makes sense.

To your earlier point, I have the following use case:

I'm authenticating/authorizing with Supabase and Firebase, using both client libraries.
There a number of steps to fully creating a proper user of a platform, and if at any point something fails, the idea was to delete the Supabase and Firebase user. Not sure what would be a better way to go about this. Cascade deletion is not set up, but expect I'll need to do that soon (hoping its in the docs somewhere). What do you think about the following [solution](https://github.com/supabase/supabase/discussions/1066#discussioncomment-1601641)? Do you think supabase.auth.admin.deleteUser() and the cascade solution in the link provides a sufficient solution to my use-case? Or what else am I overlooking?
